### PR TITLE
Nette\SmartObject & StaticClass

### PR DESCRIPTION
**Pull request by dg**
https://forum.nette.org/cs/25958-jak-se-vyporadat-s-nette-object (Czech)

### CODE REVIEW

File: src/Utils/Strict.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+		$items = array_diff($rc->getProperties(\ReflectionProperty::IS_PUBLIC), $rc->getProperties(\ReflectionProperty::IS_STATIC));
+		$hint = ($t = ObjectMixin::getSuggestion($items, $name)) ? ", did you mean $$t?" : '.';
+		throw new MemberAccessException("Attempt to write to undeclared property {$rc->getName()}::$$name$hint");
+	}
+
+
+	/**
+	 * @return bool
+	 */
+	public function __isset($name)
```
Comment (fprochazka) [inline src/Utils/Strict.php]: I was thinking about this earlier when I was [making this](https://github.com/Kdyby/StrictObjects/blob/b334bea700510844f6de31d2c2f6030f9ea3c26d/src/Scream.php#L89-L101) and I was unable to come up with benefit of returning `FALSE` here.
In Latte, it might make sense, since there are magic properties on the template object, but it does not make sense generaly imho.

If you're beeing trully strict, this potentially hides the fact, that you're doing an `isset()` check on property that is not defined.
Also, you probably never wanna do isset checks on classes that are strict, since you simply expect the property be in the definition.

Comment (dg) [inline src/Utils/Strict.php]: You are right that `return FALSE` is useless a method can be removed. I'll remove it.

File: src/Utils/Strict.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+{
+
+	/**
+	 * Call to undefined method.
+	 * @throws MemberAccessException
+	 */
+	public function __call($name, $args)
+	{
+		$class = method_exists($this, $name) ? 'parent' : get_class($this);
+		$items = (new \ReflectionClass($this))->getMethods(\ReflectionMethod::IS_PUBLIC);
```
Comment (fprochazka) [inline src/Utils/Strict.php]: I'm aware that you can call static members as methods, but should they be suggested?

Comment (dg) [inline src/Utils/Strict.php]: Why not? You can make typo in static method called as non-static.

Comment (fprochazka) [inline src/Utils/Strict.php]: Well, that is my point. Should the static methods be suggested? Since they're static and there is no reason to call them as non-static.

File: src/Utils/StrictPlus.php [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+/**
+ * Strict class for better experience.
+ * - 'did you mean' hints
+ * - access to undeclared members throws exceptions
+ * - support for @property annotations
+ * - support for calling event handlers stored in $onEvent via onEvent()
+ * - compatible with Nette\Object
+ */
+trait StrictPlus
```
Comment (Majkl578) [inline src/Utils/StrictPlus.php]: Can anyone think of better name for this? The _plus_part implies even more strict behavior, but it's not the case, instead it adds totally non-strict syntactic sugar...

**Comment by Majkl578** [comment]
:+1: I wish this was made a separate repo/project.

File: src/Utils/Strict.php [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+		), $name);
+		throw new MemberAccessException("Call to undefined static method {$rc->getName()}::$name()" . ($hint ? ", did you mean $hint()?" : '.'));
+	}
+
+
+	/**
+	 * Access to undeclared property.
+	 * @throws MemberAccessException
+	 */
+	public function &__get($name)
```
Comment (Majkl578) [inline src/Utils/Strict.php]: No need for reference here.

File: src/Utils/StrictPlus.php [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+	/**
+	 * @return void
+	 * @throws MemberAccessException
+	 */
+	public function __unset($name)
+	{
+		$class = get_class($this);
+		if (!ObjectMixin::hasProperty($class, $name)) { // strict class
+			throw new MemberAccessException("Cannot unset the property $class::\$$name.");
+		}
```
Comment (Majkl578) [inline src/Utils/StrictPlus.php]: This seems to pretend that property was sucessfully unset, but it was not, getter on such property would still work, wouldn't it?

File: src/Utils/StrictPlus.php (hunk 3 above)
Comment (dg) [inline src/Utils/StrictPlus.php]: StrictSugar?

File: src/Utils/StrictPlus.php (hunk 5 above)
Comment (dg) [inline src/Utils/StrictPlus.php]: Getter will fail https://github.com/nette/utils/blob/cf3bb5e487e57776103828d6b5ad2a446fcdf17d/tests/Utils/Object.unsetProperty.phpt#L40-L43

File: src/Utils/StrictPlus.php (hunk 3 above)
Comment (dg) [inline src/Utils/StrictPlus.php]: …or to not complicate it and have only one trait…

Comment (matej21) [inline src/Utils/StrictPlus.php]: BlackMagic? :P

Comment (JanTvrdik) [inline src/Utils/StrictPlus.php]: Enchantments? StrictEnhanced? StrictButComfortable?

Comment (fprochazka) [inline src/Utils/StrictPlus.php]: Enchantments :+1:

Comment (pavelkouril) [inline src/Utils/StrictPlus.php]: Enchantments sounds like something from a fantasy game and isn't really describing what it does. :(

Comment (JanTvrdik) [inline src/Utils/StrictPlus.php]: > **Enchantments** sounds like something from a fantasy game and isn't really describing what it does. 

I think it sort-of does. It casts a couple of spells on the given class to alter it's behavior. It's similar to BlackMagic but sounds more optimistic.

File: src/Utils/ObjectMixin.php [hunk 6]
```diff
@@ -317,6 +306,45 @@ public static function getMagicMethods($class)
...
+		foreach ($matches as list(, $type, $name)) {
+			$uname = ucfirst($name);
+			$write = (!$type || $type === '-write')
+				&& $rc->hasMethod($nm = 'set' . $uname)
+				&& ($rm = $rc->getMethod($nm)) && $rm->getName() === $nm && !$rm->isPrivate() && !$rm->isStatic();
+			$read = (!$type || $type === '-read')
+				&& ($rc->hasMethod($nm = 'get' . $uname) || $rc->hasMethod($nm = 'is' . $uname))
+				&& ($rm = $rc->getMethod($nm)) && $rm->getName() === $nm && !$rm->isPrivate() && !$rm->isStatic();
+			if ($read || $write) {
+				$props[$name] = [$read ? $nm : FALSE, $write, $read ? $rm->returnsReference() : NULL];
```
Comment (JanTvrdik) [inline src/Utils/ObjectMixin.php]: * Why store method name for getter and only boolean for setter?
*  `$read ? $rm->returnsReference() : NULL` could be simplified to `$read && $rm->returnsReference()`

Comment (JanTvrdik) [inline src/Utils/ObjectMixin.php]: If we would store only boolean for getter as well, we may replace the 2-level array with a single integer (used as bit-flags storage) for each property, which should result in lower memory footprint and faster code.

~~~php
$ref = $read && $rm->returnsReference();
$props[$name] = ($read << 0) | ($write << 1) || ($red << 2);
~~~

File: src/Utils/ObjectMixin.php [hunk 7]
```diff
@@ -25,14 +27,8 @@ class ObjectMixin
...
-
-	/**
-	 * Static class - cannot be instantiated.
-	 */
-	final public function __construct()
-	{
-		throw new Nette\StaticClassException;
-	}
+	/** @var array (name => [read name, write?, returnsByRef?])  used by getMagicProperties() */
+	private static $properties = [];
```
Comment (JanTvrdik) [inline src/Utils/ObjectMixin.php]: Having both `$props` and `$properties` is bad.

File: src/Utils/ObjectMixin.php [hunk 8]
```diff
@@ -317,6 +306,45 @@ public static function getMagicMethods($class)
 
 
 	/**
+	 * Returns array of magic properties defined by annotation @property.
+	 * @return array
+	 */
+	public static function getMagicProperties($class)
```
Comment (JanTvrdik) [inline src/Utils/ObjectMixin.php]: But on how this method is actually used, having `getMagicProperty($class, $name)` (likely returning either NULL or 0 if such property does not exist) may be reasonable.

File: src/Utils/ObjectMixin.php (hunk 6 above)
Comment (dg) [inline src/Utils/ObjectMixin.php]: Setter is always `setAbc`, getter is `getAbc` or `isAbc`.

File: src/Utils/ObjectMixin.php (hunk 7 above)
Comment (dg) [inline src/Utils/ObjectMixin.php]: Yes, but $props will be removed in future.

File: src/Utils/StrictPlus.php (hunk 3 above)
Comment (dg) [inline src/Utils/StrictPlus.php]: Comfortable sound good.

File: src/Utils/StrictPlus.php [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+		$isProp = ObjectMixin::hasProperty($class, $name);
+
+		if ($name === '') {
+			throw new MemberAccessException("Call to class '$class' method without name.");
+
+		} elseif ($isProp && $this->$name instanceof \Closure) { // closure in property
+			trigger_error("Invoking closure in property via \$obj->$name() is deprecated" . ObjectMixin::getSource(), E_USER_DEPRECATED);
+			return call_user_func_array($this->$name, $args);
+
+		} elseif ($isProp === 'event') { // calling event handlers
```
Comment (JanTvrdik) [inline src/Utils/StrictPlus.php]: Is "calling event handlers" intentionally handled after "closure in property"? Is the following behavior intentional?

~~~php
$obj->onUpdate = function () { ... };
$obj->onUpdate() // this will call the closure instead of throwing error that property `$onUpdate` must be "array or NULL".
~~~

Comment (dg) [inline src/Utils/StrictPlus.php]: Good point. This is primarily problem in Nette\Object, because StrictPlus will throw warning, although with a different message.

File: src/Utils/ObjectMixin.php (hunk 6 above)
Comment (JanTvrdik) [inline src/Utils/ObjectMixin.php]: Good point, that would require another bit flag. The `__get` initial calling sequence would then look as

~~~php
$class = get_class($this);
$uname = ucfirst($name);
$prop = ObjectMixin::getMagicProperty($class, $name);

if ($name === '') {
    throw new MemberAccessException("Cannot read a class '$class' property without name.");

} elseif ($prop & BIT_MAGIC_A) {
    return $this->{'get' . $uname}();

} elseif ($prop & BIT_MAGIC_B) {
    return $this->{'is' . $uname}();

} elseif ($prop & BIT_MAGIC_C) {
    $value = $this->{'get' . $uname}();
    return $value;

} elseif ($prop & BIT_MAGIC_D) {
    $value = $this->{'is' . $uname}();
    return $value;
}
~~~

Yeah, I know, over-optimization but that's just how I like it =)

File: src/Utils/ObjectMixin.php (hunk 7 above)
Comment (JanTvrdik) [inline src/Utils/ObjectMixin.php]: How would you implement events without `$props`? You can rename in the future the property to `$events` but you still need to store somewhere the information from reflection check.

File: src/Utils/ObjectMixin.php [hunk 10]
```diff
@@ -317,6 +306,45 @@ public static function getMagicMethods($class)
...
+			\$ (\w+)
+		()~mx', (string) $rc->getDocComment(), $matches, PREG_SET_ORDER);
+
+		$props = [];
+		foreach ($matches as list(, $type, $name)) {
+			$uname = ucfirst($name);
+			$write = (!$type || $type === '-write')
+				&& $rc->hasMethod($nm = 'set' . $uname)
+				&& ($rm = $rc->getMethod($nm)) && $rm->getName() === $nm && !$rm->isPrivate() && !$rm->isStatic();
+			$read = (!$type || $type === '-read')
```
Comment (JanTvrdik) [inline src/Utils/ObjectMixin.php]: could be simplified to `$type !== '-write'`

File: src/Utils/ObjectMixin.php [hunk 11]
```diff
@@ -317,6 +306,45 @@ public static function getMagicMethods($class)
...
+		preg_match_all('~^
+			[ \t*]*  @property(|-read|-write)  [ \t]+
+			(?: [^\s]+  [ \t]+ )?
+			\$ (\w+)
+		()~mx', (string) $rc->getDocComment(), $matches, PREG_SET_ORDER);
+
+		$props = [];
+		foreach ($matches as list(, $type, $name)) {
+			$uname = ucfirst($name);
+			$write = (!$type || $type === '-write')
```
Comment (JanTvrdik) [inline src/Utils/ObjectMixin.php]: could be simplified to `$type !== '-read'`

File: src/Utils/ObjectMixin.php (hunk 10 above)
Comment (dg) [inline src/Utils/ObjectMixin.php]: :+1:

File: src/Utils/ObjectMixin.php (hunk 6 above)
Comment (dg) [inline src/Utils/ObjectMixin.php]: It means a lot of conditions for every property usage.…

File: src/Utils/ObjectMixin.php [hunk 12]
```diff
@@ -317,6 +306,45 @@ public static function getMagicMethods($class)
...
+				&& $rc->hasMethod($nm = 'set' . $uname)
+				&& ($rm = $rc->getMethod($nm)) && $rm->getName() === $nm && !$rm->isPrivate() && !$rm->isStatic();
```
Comment (JanTvrdik) [inline src/Utils/ObjectMixin.php]: Do you intentionally return accept protected methods as well?

Comment (dg) [inline src/Utils/ObjectMixin.php]: Yes, it is feature.

File: src/Utils/ObjectMixin.php (hunk 6 above)
Comment (JanTvrdik) [inline src/Utils/ObjectMixin.php]: This first only will almost always match (90+ % by guess), the second one will handle the most of the remaining cases (7 % by guess). Getters returning references (e.g. `& getName()` or `& isValid()`) accessed as properties  are extremely rare.

File: src/Utils/ObjectMixin.php (hunk 7 above)
Comment (dg) [inline src/Utils/ObjectMixin.php]: I'll probably change it to static variable.

File: src/Utils/ObjectMixin.php (hunk 6 above)
Comment (dg) [inline src/Utils/ObjectMixin.php]: I tried it here https://github.com/nette/utils/pull/105/commits/c369f2269adda3d8c6eec4a59ff1dc49b0b01aa5

File: src/Utils/StrictPlus.php [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+				throw new MemberAccessException("Cannot read a write-only property $class::\$$name.");
+			}
+			$m = ($prop & 0b0010 ? 'get' : 'is') . $name;
+			if ($prop & 0b0100) {
+				return $this->$m();
+			} else {
+				$val = $this->$m();
+				return $val;
+			}
+		} else {
```
Comment (JanTvrdik) [inline src/Utils/StrictPlus.php]: What about

~~~php
$class = get_class($this);
$prop = ObjectMixin::getMagicProperties($class)[$name] ?? NULL;

if ($prop & 0b0001) {
    $m = ($prop & 0b0010 ? 'get' : 'is') . $name;
    if ($prop & 0b0100) {
        return $this->$m();
    } else {
        $val = $this->$m();
        return $val;
    }

} elseif ($prop !== NULL) {
    throw new MemberAccessException("Cannot read a write-only property $class::\$$name.");

} else {
    ObjectMixin::strictGet($class, $name);
}
~~~

File: src/Utils/ObjectMixin.php [hunk 14]
```diff
@@ -461,13 +502,64 @@ public static function getSuggestion(array $items, $value)
...
+		return $prop;
+	}
+
+
+	/**
+	 * Returns array of public (static, non-static and magic) methods.
+	 * @return array
+	 * @internal
+	 */
+	public static function & getMethods($class)
```
Comment (JanTvrdik) [inline src/Utils/ObjectMixin.php]: Looks like this could be private.

Comment (dg) [inline src/Utils/ObjectMixin.php]: This https://github.com/nette/utils/pull/105/commits/d465062927b7c5cc2a71d0131572338caa125fc5 will not be merged, it is only for idea.

Comment (JanTvrdik) [inline src/Utils/ObjectMixin.php]: I know but it's difficult to comment on code just before this commit.

Comment (dg) [inline src/Utils/ObjectMixin.php]: Ok, removed

File: src/Utils/Strict.php (hunk 4 above)
Comment (JanTvrdik) [inline src/Utils/Strict.php]: btw: Nette has inconsistent coding style of writing `&` before methods names, it would be nice to unify it.

File: tests/Utils/ObjectMixin.getMagicProperites().phpt [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+ * @property $cAse
+ * @property $private
+ * @property $protected
+ * @property $static
+ * @property-read $read
+ * @property-write $write
+ * @property-x $invalid
+ * abc @property $invalid2
+ */
+class TestClass
```
Comment (JanTvrdik) [inline tests/Utils/ObjectMixin.getMagicProperites().phpt]: * You test property annotations without types (except for one case), even though in the wild types are almost always present. 
* We may even consider not supporting property annotations without types (`mixed` can be used if type is unknown) to encourage programmers to write the annotations properly.
* You're missing edge case when `$this` is used as a type.

Comment (JanTvrdik) [inline tests/Utils/ObjectMixin.getMagicProperites().phpt]: * You test property annotations without types (except for one case), even though in the wild types are almost always present. 
* We may even consider not supporting property annotations without types (`mixed` can be used if type is unknown) to encourage programmers to write the annotations properly.
* You're missing edge case when `$this` is used as a type.

File: src/Utils/StrictPlus.php [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+				$rp->setValue($this, $args[0]);
+			} elseif ($op === 'add') {
+				$val = $rp->getValue($this);
+				$val[] = $args[0];
+				$rp->setValue($this, $val);
+			}
+			return $this;
+
+		} elseif ($cb = ObjectMixin::getExtensionMethod($class, $name)) { // extension methods
+			trigger_error("Extension methods like $class::$name() are deprecated" . ObjectMixin::getSource(), E_USER_DEPRECATED);
```
Comment (JanTvrdik) [inline src/Utils/StrictPlus.php]: I think this should say "Extension methods such as" instead of "Extension methods like".

File: src/Utils/StrictPlus.php [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+	 */
+	public static function extensionMethod($name, $callback = NULL)
+	{
+		if (strpos($name, '::') === FALSE) {
+			$class = get_called_class();
+		} else {
+			list($class, $name) = explode('::', $name);
+			$class = (new \ReflectionClass($class))->getName();
+		}
+		trigger_error("Extension methods like $class::$name() are deprecated" . ObjectMixin::getSource(), E_USER_DEPRECATED);
```
Comment (JanTvrdik) [inline src/Utils/StrictPlus.php]: I think this should say "Extension methods such as" instead of "Extension methods like".

Comment (JanTvrdik) [inline src/Utils/StrictPlus.php]: I think this should say "Extension methods such as" instead of "Extension methods like".

File: tests/Utils/ObjectMixin.getMagicProperites().phpt (hunk 15 above)
Comment (dg) [inline tests/Utils/ObjectMixin.getMagicProperites().phpt]: ad 1) one test is enough, isn't it?
ad 2) hmmm, yes, it can be loosen always
ad 3) I dont like to support this

File: src/Utils/StrictPlus.php (hunk 17 above)
Comment (dg) [inline src/Utils/StrictPlus.php]: :+1:

File: src/Utils/ObjectMixin.php [hunk 18]
```diff
@@ -271,6 +271,58 @@ public static function has($_this, $name)
 	}
 
 
+	/********************* magic @properties ****************d*g**/
+
+
+	/**
+	 * Returns array of magic properties defined by annotation @property.
+	 * @return array of [name => [getter name?, allow write?, returns by ref?]]
```
Comment (JanTvrdik) [inline src/Utils/ObjectMixin.php]: phpDoc not updated

File: src/Utils/ObjectMixin.php [hunk 19]
```diff
@@ -270,6 +263,57 @@ public static function has($_this, $name)
...
+		}
+		return $props;
+	}
+
+
+	/** @internal */
+	public static function getMagicProperty($class, $name)
+	{
+		$props = self::getMagicProperties($class);
+		return isset($props[$name]) ? $props[$name] : NULL;
```
Comment (JanTvrdik) [inline src/Utils/ObjectMixin.php]: I still think that 0 is slightly better than NULL (you can declare return type as int, you avoid casting NULL to integer when using bit-wise and to check some bit, it is shorter)

**Comment by hrach** [comment]
I wonder, wouldn't be better to push traits into Utils subnamespace? Now it's the ideal time.

**Comment by enumag** [comment]
More like into a separate package in my opinion.

**Comment by dg** [comment]
I do not want to force all users Nette and authors of addons to require a new package.

**Comment by hrach** [comment]
I understand that. That's why I suggest putting this into correct package namespace. :)

**Comment by dg** [comment]
I am rather surprised that nobody did not address namespace :-)  https://forum.nette.org/cs/25958-jak-se-vyporadat-s-nette-object#p172395

I have no idea.

**Comment by JanTvrdik** [comment]
> wouldn't be better to push traits into Utils subnamespace?

Good point. I agree, now is the ideal time to do it.

> More like into a separate package in my opinion.

I disagree, nette/utils is imho the ideal place.

> I do not want to force all users Nette and authors of addons to require a new package.

nette/utils would depend on the new package, so no manual change would be required by Nette users

**Comment by dg** [comment]
> nette/utils would depend on the new package, so no manual change would be required by Nette users

Not so easy, because *the new package* depends on nette\utils too and I dont want to think two days how to solve chicken egg problem.

**Comment by Majkl578** [comment]
It's sad to see the wishes to have a StrictObject trait were ignored... Many people were hoping to have a trait with strictness, but without enforced magic. :disappointed:

**Comment by fprochazka** [comment]
Cough, cough, [already exists](https://github.com/Kdyby/StrictObjects), cough, cough :smirk:

**Comment by castamir** [comment]
@fprochazka cool trait, terrible name :D

**Comment by dg** [comment]
StrictObject is not needed.
